### PR TITLE
improve URL

### PR DESCRIPTION
**Pull request by valq7711**
allow to map domains to apps
example of usage:
```nginx
 map $http_host $appname {
       hostnames;
       foo.*   /some;  # foo.some.org  mapped to '/some' app
       .bar.com   /other;
 }
 server {
        listen       80;
        listen       [::]:80;
        server_name  foo.baz.org  www.bar.com  .example.*;

    location / {
        proxy_set_header   X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header   Host $http_host;
        proxy_set_header   X-Real-IP $remote_addr;
        proxy_set_header   X-Forwarded-Host $server_name;
        proxy_redirect off;
        # for app that is not mapped we should just unset X-Py4web-Appname
        location /not_mapped_app/ {
                proxy_set_header X-Py4web-Appname "";
                proxy_pass http://unix:/run/gunicorn.sock;
        }
       # if this header is set and not empty then app is mapped => URL() should work with use_appname= False
        proxy_set_header X-Py4web-Appname $appname;

        # magic here  - we just add $appname to uri 
        # note that .example.* is not mapped and in this case $appname will be ""
        # so for  example.com  all apps  will be served in regular mode! i.e. as http://example.com/some 
        rewrite ^ $appname$request_uri break;
        proxy_pass http://unix:/run/gunicorn.sock;
    }
}


```

### CODE REVIEW

File: py4web/core.py [hunk 1]
```diff
@@ -591,6 +591,9 @@ def URL(
     URL('a','b',vars=dict(x=1),scheme='https') -> https://{domain}/{script_name?}/{app_name}/a/b?x=1
     URL('a','b',vars=dict(x=1),use_appname=False) -> /{script_name?}/a/b?x=1
     """
+    if use_appname is None:
+        use_appname = request.headers.get("x-py4web-appname")
+        use_appname = True if use_appname is None else not use_appname
```
Comment (mdipierro) [inline py4web/core.py:596]: Do not understand this. "x-py4web-appname": "false" would evaluate to true?

Comment (valq7711) [reply py4web/core.py:596]: Yes, it looks horrible and would be better to have `proxy_set_header x-py4web-use-appname !$appname`, but nginx doesnt support these things and it requires if-block + yet another var

Comment (valq7711) [reply py4web/core.py:596]: Just in case:
https://www.nginx.com/resources/wiki/start/topics/depth/ifisevil/

**Comment by misl6** [comment]
For being sure that this one works out-of-the-box also for apps that uses the auth module, how it sounds to change also this param to be `None` by default?

https://github.com/web2py/py4web/blob/c1dcf19a837a3835684d8f1e86c051a500499605/py4web/utils/auth.py#L145

**Comment by valq7711** [comment]
@mdipierro    the logic is absolutely the same as for x-script-name: the proxy server adds a prefix to the uri and passes information about this to wsgi

**Comment by valq7711** [comment]
@mdipierro  Why is `use_appname` needed at all?

**Pull request merged**